File: src/main/java/me/kuku/yuq/controller/HostLocController.java
package me.kuku.yuq.controller;

import com.IceCreamQAQ.Yu.annotation.Action;
import com.IceCreamQAQ.Yu.annotation.Before;
import com.IceCreamQAQ.Yu.annotation.Synonym;
import com.icecreamqaq.yuq.annotation.GroupController;
import com.icecreamqaq.yuq.annotation.PathVar;
import com.icecreamqaq.yuq.annotation.PrivateController;
import com.icecreamqaq.yuq.annotation.QMsg;
import com.icecreamqaq.yuq.entity.Group;
import com.icecreamqaq.yuq.message.MessageItemFactory;
import me.kuku.pojo.Result;
import me.kuku.yuq.entity.HostLocEntity;
import me.kuku.yuq.entity.HostLocService;
import me.kuku.yuq.entity.QqEntity;
import me.kuku.yuq.logic.HostLocLogic;

import javax.inject.Inject;
import java.io.IOException;

@GroupController
@PrivateController
public class HostLocController {

	@Inject
	private HostLocService hostLocService;
	@Inject
	private HostLocLogic hostLocLogic;
	@Inject
	private MessageItemFactory mif;

	@Action("loc {username} {password}")
	public String login(QqEntity qqEntity, String username, String password) throws IOException {
		Result<String> result = hostLocLogic.login(username, password);
		if (result.isFailure()) return result.getMessage();
		HostLocEntity hostLocEntity = hostLocService.findByQqEntity(qqEntity);
		if (hostLocEntity == null) hostLocEntity = HostLocEntity.Companion.getInstance(qqEntity);
		hostLocEntity.setCookie(result.getData());
		hostLocService.save(hostLocEntity);
		return "绑定HostLoc成功！";
	}

	@Before(except = "login")
	public HostLocEntity before(QqEntity qqEntity){
		HostLocEntity hostLocEntity = hostLocService.findByQqEntity(qqEntity);
		if (hostLocEntity == null)
			throw mif.at(qqEntity.getQq()).plus("您还未绑定HostLoc账号，请先绑定HostLoc账号！").toThrowable();
		else return hostLocEntity;
	}

	@Action("loc签到")
	@QMsg(at = true)
	public String locSign(HostLocEntity hostLocEntity, long qq, Group group) throws IOException {
		group.sendMessage(mif.at(qq).plus("正在为您进行loc签到中，请稍后！"));
		String cookie = hostLocEntity.getCookie();
		boolean isLogin = hostLocLogic.isLogin(cookie);
		if (isLogin) hostLocLogic.sign(cookie);
		else {
			return "loc签到失败，cookie失效，请重新登录！";
		}
		return "loc签到成功！";
	}

	@Action("loc签到 {status}")
	@Synonym({"loc监控 {status}"})
	public String status(HostLocEntity hostLocEntity, boolean status, @PathVar(0) String type){
		switch (type){
			case "loc监控": hostLocEntity.setMonitor(status); break;
			case "loc签到": hostLocEntity.setSign(status); break;
		}
		hostLocService.save(hostLocEntity);
		return type + (status ? "开启成功" : "关闭成功");
	}

}

File: src/main/kotlin/me/kuku/yuq/controller/HostLocController.kt
package me.kuku.yuq.controller

import com.IceCreamQAQ.Yu.annotation.Action
import com.IceCreamQAQ.Yu.annotation.Before
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.annotation.PrivateController
import com.icecreamqaq.yuq.controller.BotActionContext
import com.icecreamqaq.yuq.controller.ContextSession
import com.icecreamqaq.yuq.message.Message.Companion.firstString
import com.icecreamqaq.yuq.mif
import me.kuku.yuq.entity.HostLocEntity
import me.kuku.yuq.entity.HostLocService
import me.kuku.yuq.entity.QqEntity
import me.kuku.yuq.logic.HostLocLogic
import me.kuku.yuq.utils.openOrClose
import org.springframework.stereotype.Component

@GroupController
@PrivateController
@Component
class HostLocController (
    private val hostLocService: HostLocService
) {

    @Action("loc登录")
    suspend fun locLogin(qqEntity: QqEntity, session: ContextSession, context: BotActionContext): String {
        context.source.sendMessage(mif.at(qqEntity.qq).plus("请发送账号").toMessage())
        val account = session.waitNextMessage().firstString()
        context.source.sendMessage(mif.at(qqEntity.qq).plus("请发送密码").toMessage())
        val password = session.waitNextMessage().firstString()
        val res = HostLocLogic.login(account, password)
        return if (res.success()) {
            val cookie = res.data()
            val hostLocEntity = hostLocService.findByQqEntity(qqEntity) ?: HostLocEntity().also {
                it.qqEntity = qqEntity
            }
            hostLocEntity.cookie = cookie
            hostLocService.save(hostLocEntity)
            "绑定HostLoc成功"
        } else res.message
    }

    @Before(except = ["locLogin"])
    fun before(qqEntity: QqEntity, qq: Long): HostLocEntity {
        return hostLocService.findByQqEntity(qqEntity) ?: throw mif.at(qq).plus("你未绑定HostLoc，签到失败").toThrowable()
    }

    @Action("loc签到")
    fun locSign(hostLocEntity: HostLocEntity): String {
        val cookie = hostLocEntity.cookie
        HostLocLogic.sign(cookie)
        return "loc签到成功"
    }

    @Action("loc推送 {status}")
    fun push(status: Boolean, hostLocEntity: HostLocEntity): String {
        hostLocEntity.config.push = status.toStatus()
        hostLocService.save(hostLocEntity)
        return "loc推送${if (status) "开启" else "关闭"}成功"
    }

    @Action("loc自动签到 {status}")
    fun locAutoSign(hostLocEntity: HostLocEntity, status: Boolean): String {
        hostLocEntity.config.sign = status.toStatus()
        hostLocService.save(hostLocEntity)
        return "loc自动签到${status.openOrClose()}成功"
    }

}